### PR TITLE
fix(model_init): stream Devstral FP8 checkpoints

### LINKED ISSUE #2114: FineGrainedFP8Config(dequantize=true) materializes full BF16 model per rank before sharding
## What

The standard FP8-checkpoint load path

\`\`\`yaml
quantization_config:
  _target_: transformers.FineGrainedFP8Config
  dequantize: true
\`\`\`

dequantizes the **entire** FP8 checkpoint into BF16 **on every rank** before any TP / FSDP / PP sharding kicks in. The full BF16 model is the per-rank peak.

This is fine on small models (24B BF16 ≈ 48 GB, fits 80 GB H100) but OOMs at scale.

## Why it matters

| Model | Full BF16 footprint | Fits 80 GB H100? | Fits 141 GB H200? |
|---|---|---|---|
| Devstral-Small-2-24B (FP8) | ~48 GB | yes | yes |
| Devstral-2-123B (FP8) | ~246 GB | no | no |
| Mistral-Medium-3.5-128B (FP8) | ~256 GB | no | no |

For 123B / 128B+ FP8 checkpoints there's *no* per-rank configuration — even TP=8 PP=8 across 64 GPUs — that lets the HF dequant path run, because the OOM happens *before* the model is sharded. Cranking up node count does nothing.

## Today's solution

\`nemo_automodel/components/models/mistral3_vlm/state_dict_adapter.py\` (see \`Mistral3FP8StateDictAdapter.for_vlm_full()\`) hooks the DCP load so each rank only reads its own shard from disk: pull the FP8 weight + per-tensor \`weight_scale_inv\` for the local shard, dequantize to BF16 in place, emit the local DTensor. Per-rank peak ≈ shard size, not full model.

This works but is a one-off — it lives inside the \`mistral3_vlm\` package and only kicks in when the resolver hook claims the config (i.e. for the dawn-ridge 128B specifically). Any other FP8-native dense / MoE checkpoint at this scale would need to ship its own adapter.

Possible shapes:
1. Generalize the \`Mistral3FP8StateDictAdapter\` factory pattern into a \`StreamingFP8StateDictAdapter\` keyed off \`config.quantization_config.quant_method == \"fp8\"\`, registered for any model whose checkpoint declares per-tensor or per-block FP8.
2. Push the streaming dequant upstream into \`transformers/quantizers/quantizer_finegrained_fp8.py\` (the loader that processes \`FineGrainedFP8Config\`) so HF itself reads + dequants per shard rather than per full tensor.
3. Document the failure mode loudly: today the OOM happens deep in the HF loader with no log line saying \"this can't work at this scale, use a streaming adapter\". A pre-flight check in \`nemo_automodel/_transformers/model_init.py\` that estimates BF16 footprint vs available HBM and refuses to start would at least convert the silent OOM into a useful error.

## Repro

Any FP8 checkpoint > ~80 GB BF16 with \`FineGrainedFP8Config(dequantize=true)\`, e.g. \`mistralai/Devstral-2-123B-Instruct-2512\` on 8 × H100-80GB:

\`\`\`yaml
model:
  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
  pretrained_model_name_or_path: mistralai/Devstral-2-123B-Instruct-2512
  force_hf: true
  quantization_config:
    _target_: transformers.FineGrainedFP8Config
    dequantize: true
\`\`\`

OOMs at HF dequant before any sharding.

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
## Summary

* add shard-aware per-tensor FP8 loading for text-only `Ministral3ForCausalLM` checkpoints such as `mistralai/Devstral-2-123B-Instruct-2512`
* move the existing Mistral3 FP8 state-dict adapter to the shared Mistral3 family package, preserve the legacy VLM import path, and add an identity-layout causal-LM factory
* keep per-block FP8 explicitly unsupported instead of applying the scalar-scale formula incorrectly
* make the FP8 pre-flight message route-aware and point supported users to the custom streaming path
* route `ministral3` checkpoints through the offline-aware Mistral tokenizer backend
* add a TP=8 LoRA example recipe for Devstral-2 123B

## Motivation

The Transformers FP8 dequantization route materializes the full BF16 model on every rank before Automodel sharding. A 123B checkpoint therefore cannot be rescued by tensor parallelism on that route.

Automodel already has model-owned, shard-aware adapters for FP8 families. The missing piece for this checkpoint is narrower: the local text-only `Ministral3ForCausalLM` did not attach the existing Mistral3 per-tensor FP8 adapter.

This change intentionally stays family-scoped. The existing FP8 models have different key layouts, scale metadata, tensor transforms, and sharding contracts, so a generic adapter selected only by `quant_method == "fp8"` would hide model-specific assumptions that are not yet uniform. We can generalize later when multiple adapters share a proven contract.

Addresses NVIDIA-NeMo/Automodel#2114.

## Validation

* post-rebase focused guard, FP8/PEFT, Mistral3 text/VLM adapter, resolver, and optimized TP suite: 172 passed
* focused tokenizer suite: 72 passed
* Ruff format and lint checks pass for all changed Python files
* TP=2 + FSDP2 + LoRA validation on two H100s:
  * all base parameters exactly matched independent FP8 × BF16-scale dequantization
  * 16 parameters were truly sharded DTensors
  * LoRA parameters were trainable
  * finite forward logits
* official Devstral-2 123B checkpoint metadata matches the supported contract: FP8 E4M3 linear weights with scalar BF16 `weight_scale_inv`
* [scoped release CI](<https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/61327136>) passed on the exact PR commit with the release template's `HF_HUB_OFFLINE=1` unchanged:
  * loaded the cached checkpoint and dequantized 616 FP8 weights
  * selected `MistralCommonBackend` for checkpoint model type `ministral3`
  * completed five TP=8 + LoRA training steps with finite loss and gradient norms inside the 30-minute limit

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
/ok to test [a2f28a3](https://github.com/NVIDIA-NeMo/Automodel/pull/3413/commits/a2f28a36da9300ada82abef9ad699a08ea3de67c)

**Comment by yuhezhang-ai** [comment]
/ok to test [85b0d90](https://github.com/NVIDIA-NeMo/Automodel/pull/3413/commits/85b0d9095031040beaebd4bedf6577a44d5aaaff)

**Comment by yuhezhang-ai** [comment]
/ok to test [3a7abd6](https://github.com/NVIDIA-NeMo/Automodel/pull/3413/commits/3a7abd67288cbd2457088922135dc4c65330fe35)

**Comment by yuhezhang-ai** [comment]
@HuiyingLi  can you help review this pr?
I didn't include it for release, so no hurry.

**Review by akoumpa — APPROVED** [review]
